plugins/src/gcc/imp.rs

```diff
@@ -0,0 +1,1270 @@
+                        reason: format!("Over use detected {:#?}", self.detector),
+                    };
+                    self.ema.update(effective_bitrate);
+                    self.last_decrease_on_delay = now;
+
+                    return self.set_bitrate(bwe, target as Bitrate, ControllerType::Delay);
+                }
+            }
+            NetworkUsage::Under => {
+                if let BandwidthEstimationOp::Increase = self.last_control_op {
+                    if let Some(bitrate) = self.compute_increased_rate(bwe) {
+                        return self.set_bitrate(bwe, bitrate, ControllerType::Delay);
+                    }
+                }
+            }
+        }
+
+        self.last_control_op = BandwidthEstimationOp::Hold;
+
+        false
+    }
+}
+
+pub struct BandwidthEstimator {
+    state: Mutex<State>,
+
+    srcpad: gst::Pad,
+    sinkpad: gst::Pad,
+}
+
+impl BandwidthEstimator {
+    fn push_list(&self, list: gst::BufferList) {
+        let res = self.srcpad.push_list(list);
+
+        self.state.lock().unwrap().flow_return = res;
+    }
+
+    fn start_task(&self, bwe: &super::BandwidthEstimator) -> Result<(), glib::BoolError> {
+        let weak_controller = bwe.downgrade();
+        let weak_pad = self.srcpad.downgrade();
+        let clock = gst::SystemClock::obtain();
+
+        bwe.imp().state.lock().unwrap().periodic_push = Some(clock.new_periodic_id(
+            clock.time().unwrap() + dur2ts(*BURST_TIME),
+            dur2ts(*BURST_TIME),
+        ));
+
+        self.srcpad.start_task(move || {
+            let pause = || {
+                if let Some(pad) = weak_pad.upgrade() {
+                    pad.pause_task().unwrap();
+                }
+            };
+            let bwe = weak_controller
+                .upgrade()
+                .expect("bwe destroyed while its srcpad task is still running?");
+
+            let lock_state = || bwe.imp().state.lock().unwrap();
+
+            let periodic_id = match lock_state().periodic_push.take() {
+                Some(id) => id,
+                _ => {
+                    gst::info!(CAT, "Peridic pushes are stop, pausing task");
+                    return pause();
+                }
+            };
+
+            if let (Err(err), _) = periodic_id.wait() {
+                match err {
+                    gst::ClockError::Early => (),
+                    _ => {
+                        gst::error!(CAT, "Got error {err:?} on the clock, pausing task");
+
+                        lock_state().flow_return = Err(gst::FlowError::Flushing);
+
+                        return pause();
+                    }
+                }
+            }
+            let list = {
+                let mut state = lock_state();
+                state.periodic_push = Some(periodic_id);
+                state.create_buffer_list(&bwe)
+            };
+
+            if list.len() > 0 {
+                let _ = bwe.imp().push_list(list);
```

Comment:
No need for error handling here, not even logging?

Reply:
@thiblahute ?

Reply:
Look at the current code?

``` rs
    fn push_list(&self, list: gst::BufferList) {
        let res = self.srcpad.push_list(list);

        self.state.lock().unwrap().flow_return = res;
    }
```

this was solved for a good reason.

Reply:
It wasn't solved @thiblahute !

Reply:
also you're now assigning "void" to `_`. Can you explain why you don't directly post error messages instead of relying on upstream failing to chain to error out ooc?

Reply:
WTF github hides unsolved conversations... oO

I took example on gstqueue to let upstream handle the flow error, but I didn't notice that queue also pauses the task when that happens. It is now done that way, meaning that `self.push_list` now returns the value and in the GstTask callback I pause and return when downstream returns an error.